internal/controller/xlinecluster_controller.go

```diff
@@ -36,27 +37,66 @@ type XlineClusterReconciler struct {
 //+kubebuilder:rbac:groups=xline.kvstore.datenlord.com,resources=xlineclusters,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=xline.kvstore.datenlord.com,resources=xlineclusters/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=xline.kvstore.datenlord.com,resources=xlineclusters/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the XlineCluster object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/reconcile
 func (r *XlineClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	recCtx := reconciler.NewReconcileContext(r.Client, r.Scheme, ctx)
```

Comment:
I found a package that can run multiple reconciling processes in parallel. https://github.com/risingwavelabs/ctrlkit. Could we use it in next version?

Reply:
The star and commit counts for this package are both limited. IMO, unless we explicitly encounter performance issues, I'm not inclined to use it in the next version.